### ISSUE TITLE
invalid double parsing in reql protocol

### ISSUE THREAD

**Issue by mbroadst** [rethinkdb/rethinkdb#5989]
In switching the C++ driver to use rapidjson I had only one failure in the numbers tests:
```
  TEST: maybe_run(R::expr(1234567890123456789012345678901234567890.0), *conn)
[11] >> [1,1.2345678901234569e39,{}]
[11] << {"r":[1.2345678901234567e39],"t":1}
  FAILURE in ‘maybe_run(R::expr(1234567890123456789012345678901234567890.0), *conn)’:
    Expected: ‘1.2345678901234569e39’
     but got: ‘1.2345678901234567e39’
```

with the original code the number is encoded as follows:
```
 TEST: maybe_run(R::expr(1234567890123456789012345678901234567890.0), *conn)
[11] >> [1,1.2345678901234568e+39,{}]
[11] << {"r":[1.2345678901234568e+39],"t":1}
```

I can confirm that `1.2345678901234568e+39` == `1.2345678901234569e39`, however when the latter is sent to the server it's responding with `1.2345678901234567e39`.

Any thoughts here?

**Comment by bchavez** [comment]
Hope/maybe this helps: https://github.com/rethinkdb/rethinkdb/issues/5349 :+1: 

:) 
----
:boom: :fire: [***"Set it ablaze like a candle wick... Light it up, light it up..."***](https://www.youtube.com/watch?v=qDcFryDXQ7U)

**Comment by mbroadst** [comment]
@bchavez thanks for the link, actually I was working on this with Etienne and he suggested the fix that went into RethinkDB at the time (enabling `-DRAPIDJSON_PARSE_DEFAULT_FLAGS=kParseFullPrecisionFlag`) which I did in the driver to the same effect, so I'm a bit stumped here.

**Comment by danielmewes** [comment]
So just to be clear, if you send the server `1.2345678901234568e+39`, it responds with the same number `1.2345678901234568e+39` (8 at the end of both). But if you send it the minimally larger (but as a double actually equal) number `1.2345678901234569e39`, (with a 9), it responds with a slightly *smaller* number `1.2345678901234567e39` (7 at the end)?

That's definitely suspicious. We should look into that.

**Comment by mbroadst** [comment]
@danielmewes exactly. what's even more curious is that this is a case where we are essentially passing rapidjson encoded data to a rapidjson parser on your side, albeit I'm using v1.0.2 and you guys are using v0.1 (though that seems unlikely to be the cause?)

**Comment by mlucy** [comment]
We've had problems in the past with JSON encoders not actually being able to serialize/deserialize a double and get the same result back.  It can cause surprisingly subtle problems, so we should definitely look into it.
